File: FriendScore.java
import java.util.*;

public class FriendScore {
  public int highestScore(String[] friends) {
    int n = friends.length;
    int[] scores = new int[n];

    for (int i = 0; i < n; i++) {
      for (int j = 0; j < n; j++) {
        if (i == j)
          continue;

        String f = friends[i];
        Set<Integer> count = new HashSet<>();

        for (int k = 0; k < n; k++) {
          if (f.charAt(k) == 'N')
            continue;

          count.add(k);

          for (int l = 0; l < n; l++) {
            if (i == l)
              continue;

            char c = friends[l].charAt(k);
            if (c == 'N')
              continue;

            count.add(l);
          }
        }

        scores[i] = count.size();
      }
    }

    int ans = 0;
    for (int score : scores) {
      ans = Math.max(ans, score);
    }

    return ans;
  }

  // public static void main(String[] args) {
  //   System.out.println(new FriendScore().highestScore(
  //       new String[] { "NNNNNNNNNNNNNNY", "NNNNNNNNNNNNNNN", "NNNNNNNYNNNNNNN", "NNNNNNNYNNNNNNY", "NNNNNNNNNNNNNNY",
  //           "NNNNNNNNYNNNNNN", "NNNNNNNNNNNNNNN", "NNYYNNNNNNNNNNN", "NNNNNYNNNNNYNNN", "NNNNNNNNNNNNNNY",
  //           "NNNNNNNNNNNNNNN", "NNNNNNNNYNNNNNN", "NNNNNNNNNNNNNNN", "NNNNNNNNNNNNNNN", "YNNYYNNNNYNNNNN" }));
  // }
}

File: friend_score.cpp
#include <string>
#include <vector>
#include <map>
#include <algorithm>
#include <functional>
#include <set>
#include <iostream>
using namespace std;

class FriendScore
{
public:
  int highestScore(vector<string> friends)
  {
    int ans = 0;
    int n = friends.size();
    for (int i = 0; i < n; i++)
    {
      ans = max(ans, score(friends, i));
    }
    return ans;
  }

private:
  int score(vector<string> friends, int i)
  {
    int n = friends.size();
    string f = friends[i];
    set<int> nums;
    for (int j = 0; j < n; j++)
    {
      if (f[j] == 'N')
      {
        continue;
      }

      nums.insert(j);
      for (int k = 0; k < n; k++)
      {
        if (k == i)
        {
          continue;
        }

        if (friends[k][j] == 'Y')
        {
          nums.insert(k);
        }
      }
    }

    return nums.size();
  }
};

int main()
{
  FriendScore *fs = new FriendScore();
  cout << fs->highestScore({"NYNNN", "YNYNN", "NYNYN", "NNYNY", "NNNYN"}) << endl;
}
